src/Sentry/Protocol/SpanRecorder.cs

```diff
@@ -0,0 +1,40 @@
+using System.Collections.Generic;
+
+namespace Sentry.Protocol
+{
+    /// <summary>
+    /// Records spans that belong to the same transaction.
+    /// </summary>
+    public class SpanRecorder
+    {
+        private const int MaxSpans = 1000;
+
+        private readonly object _lock = new object();
+        private readonly List<Span> _spans = new List<Span>();
+
+        /// <summary>
+        /// Records a span.
+        /// </summary>
+        public void Add(Span span)
+        {
+            lock (_lock)
+            {
+                if (_spans.Count < MaxSpans)
+                {
+                    _spans.Add(span);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets all recorded spans.
+        /// </summary>
+        public IReadOnlyList<Span> GetAll()
+        {
+            lock (_lock)
+            {
+                return _spans;
```

Comment:
There's exclusive access to just return the reference, which now can be access outside of any locks by the caller which can iterate over it while another thread is adding to it, which would throw an exception.

We need a new list, with the current items:

```suggestion
                return new List<Span>(_spans);
```